src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/VsTargetFrameworkInfo.cs

```diff
@@ -11,33 +11,67 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore;
 ///     interface for NuGet;
 /// </summary>
 [DebuggerDisplay("TargetFrameworkMoniker = {TargetFrameworkMoniker}")]
-internal class VsTargetFrameworkInfo : IVsTargetFrameworkInfo3
+internal class VsTargetFrameworkInfo : IVsTargetFrameworkInfo4
 {
     private readonly TargetFrameworkInfo _targetFrameworkInfo;
-    
-    private VsReferenceItems? _packageDownloads;
-    private VsReferenceItems? _frameworkReferences;
-    private VsReferenceItems? _projectReferences;
-    private VsReferenceItems? _packageReferences;
-    private VsReferenceItems? _centralPackageVersions;
-    private VsProjectProperties? _properties;
+
+    private IReadOnlyDictionary<string, string?>? _properties;
+    private IReadOnlyDictionary<string, IReadOnlyList<IVsReferenceItem2>>? _items;
 
     public VsTargetFrameworkInfo(TargetFrameworkInfo targetFrameworkInfo)
     {
         _targetFrameworkInfo = targetFrameworkInfo;
     }
 
-    public IVsReferenceItems PackageDownloads => _packageDownloads ??= new VsReferenceItems(_targetFrameworkInfo.PackageDownloads);
-
-    public IVsReferenceItems FrameworkReferences => _frameworkReferences ??= new VsReferenceItems(_targetFrameworkInfo.FrameworkReferences);
-
     public string TargetFrameworkMoniker => _targetFrameworkInfo.TargetFrameworkMoniker;
 
-    public IVsReferenceItems ProjectReferences => _projectReferences ??= new VsReferenceItems(_targetFrameworkInfo.ProjectReferences);
+    public IReadOnlyDictionary<string, string?> Properties => _properties ??= _targetFrameworkInfo.Properties!;
```

Comment:
Null forgiveness seems redundant here, if I'm following correctly in the browser:

```suggestion
    public IReadOnlyDictionary<string, string?> Properties => _properties ??= _targetFrameworkInfo.Properties;
```

Reply:
There's a compiler warning, and VS shows a green squiggly line without the `!`

> warning CS8619: Nullability of reference types in value of type 'IImmutableDictionary<string, string>' doesn't match target type 'IReadOnlyDictionary<string, string?>'.

But it is a warning, not an error. So, it's up to you if you prefer the warning or the `!`.

To be honest, I set the property value to be nullable, not because I wanted project systems to send NuGet null property values, but because I can't trust 3rd party project systems to do the right thing, and telling the compiler that the value is nullable gives me compiler warnings everywhere in NuGet's code where I didn't null check values correctly.

It'd be nice to be able to use one set of annotations for public APIs, and a different set of annotations for in-project null checks.

Reply:
If these are evaluated property values, then they can't be `null` as far as I know. Only unevaluated values can be null. Why does the value need to be nullable?

Reply:
Rephrasing what I wrote in my previous comment, VS doesn't restrict NuGet's APIs to only dotnet/project-system. If some 3rd party writes a project-system and calls the API, but passes a null value (say, they don't enable nullable checks in their repo), then by having NuGet handle nulls gracefully, we avoid accidental NullReferenceExceptions that might be hard to diagnose.

Reply:
This API is modelling MSBuild concepts (items/properties) and should use the correct types and nullabilities.

If NuGet wants to defend against 3rd parties ignoring null annotations, that's fine, but it doesn't help the robustness of our system to permit more values than needed. And these days nearly every NullReferenceException I see comes from code that suppressed a nullability warning.

In general, being overly permissive/liberal in what an API accepts leads to higher maintenance costs. Having "one way" of doing something is clearer and less confusing. If you accept null, the developer needs to ask themselves "what's the difference between `null` and `""` here?".